embedded/src/main/java/org/wildfly/core/embedded/Configuration.java

```diff
@@ -0,0 +1,421 @@
+         */
+        public Builder addSystemPackages(final String... systemPackages) {
+            if (MODULE_LOADER_CONFIGURED.get()) {
+                EmbeddedLogger.ROOT_LOGGER.moduleLoaderAlreadyConfigured(SYSPROP_KEY_SYSTEM_MODULES);
+            }
+            if (systemPackages != null) {
+                this.systemPackages.addAll(Arrays.asList(systemPackages));
+            }
+            return this;
+        }
+
+        /**
+         * Sets the system packages for the module loader.
+         * <p>
+         * Note that this value is only used if the {@linkplain #setModuleLoader(ModuleLoader) module loader} is
+         * {@code null}. Also note that once the module loader is created changing this property, even for new
+         * configurations, will have no effect in the same JVM.
+         * </p>
+         *
+         * @param systemPackages the system package to add, {@code null} clears the previously set system packages
+         *
+         * @return this builder
+         */
+        public Builder setSystemPackages(final String... systemPackages) {
+            if (MODULE_LOADER_CONFIGURED.get()) {
+                EmbeddedLogger.ROOT_LOGGER.moduleLoaderAlreadyConfigured(SYSPROP_KEY_SYSTEM_MODULES);
+            }
+            this.systemPackages.clear();
+            return addSystemPackages(systemPackages);
+        }
+
+        /**
+         * Creates a new immutable configuration.
+         *
+         * @return the configuration
+         */
+        public Configuration build() {
+            final Path jbossHome = this.jbossHome;
+            final LoggerHint loggerHint = this.loggerHint == null ? LoggerHint.DEFAULT : this.loggerHint;
+            final boolean embeddedLoggingHint = this.embeddedLoggingHint;
+            configureLogging(loggerHint);
+            final String[] cmdArgs = this.cmdArgs.toArray(new String[0]);
+            final String[] systemPackages = this.systemPackages.toArray(new String[0]);
+            final ModuleLoader moduleLoader;
+            if (this.moduleLoader == null) {
+                final String modulePath = WildFlySecurityManager.getPropertyPrivileged(SYSPROP_KEY_MODULE_PATH,
+                        jbossHome.resolve(JBOSS_MODULES_DIR_NAME).toAbsolutePath().toString());
+                moduleLoader = setupModuleLoader(modulePath, systemPackages);
+                MODULE_LOADER_CONFIGURED.set(true);
+            } else {
+                moduleLoader = this.moduleLoader;
+            }
+            return new Configuration() {
+                @Override
+                public Path getJBossHome() {
+                    return jbossHome;
+                }
+
+                @Override
+                public ModuleLoader getModuleLoader() {
+                    return moduleLoader;
+                }
+
+                @Override
+                public String[] getCommandArguments() {
+                    return Arrays.copyOf(cmdArgs, cmdArgs.length);
```

Comment:
Why are you copying the `cmdArgs` array instead of passing it directly?

Reply:
Only to ensure the `Configuration` is immutable.

Reply:
I don't think that's necessary. `cmdArgs` is already an array that is created in `build()` by calling `this.cmdArgs.toArray(new String[0]);`.